### PR TITLE
Allow GraphQL documents to be typed with a TypedDocumentNode for fully-typed result and variables objects

### PR DESCRIPTION
**Pull request by dthyresson**
Fixes https://github.com/redwoodjs/redwood/issues/9232

See the above issue for key points.

TLDR;

Before this PR, it wasn't possible to infer types from the `useQuery` or `useMutation` GraphQL hooks. 

Also, `TypedDocumentNode` was not conveniently exported from `@redwoodjs/web` (and is now).

Now, you can do:

```ts
import type { FindPostByIdVariables, FindPostById } from 'types/graphql'

import type {
  CellSuccessProps,
  CellFailureProps,
  TypedDocumentNode,
} from '@redwoodjs/web'

import Post from 'src/components/Post/Post'

export const QUERY: TypedDocumentNode<
  FindPostById,
  FindPostByIdVariables
> = gql`
  query FindPostById($id: Int!) {
    post: post(id: $id) {
      id
      title
      body
      authorId
      createdAt
    }
  }
````

and infer types:

image

image

For context:

See https://the-guild.dev/blog/typed-document-node 

> TypedDocumentNode is a development tool for creating fully typed DocumentNode objects. It means that just by passing the GraphQL query/mutation/subscription/fragment to a supporting GraphQL client library, you'll get fully-typed result and variables objects.

> This is made possible by [TypeScript type inference](https://typescriptlang.org/docs/handbook/type-inference.html).

TypedDocumentNode is a development tool that allows you to create fully typed DocumentNode objects. This means that by simply passing your GraphQL query, mutation, subscription, or fragment to a supporting GraphQL client library, you'll get fully-typed result and variables objects. This is made possible by TypeScript type inference.

### CODE REVIEW

**Comment by dthyresson** [comment]
@Tobbe and others: one question I had is if the generators should be updated to detect if using TS and infer types as above. If som, we can make this a separate PR enhancement to update the generators.

File: packages/internal/src/generate/clientPreset.ts [hunk 1]
```diff
@@ -32,27 +32,6 @@ export const generateClientPreset = async () => {
     schema: getPaths().generated.schema,
     documents: documentsGlob,
     generates: {
-      // should be graphql.d.ts
```
Comment (Tobbe) [inline packages/internal/src/generate/clientPreset.ts:35]: Why can you remove this now? How is this related to just adding `| TypedDocumentNode<TData, TVariables>` to some types in `packages/web/src/components/GraphQLHooksProvider.tsx`?

Comment (dthyresson) [reply packages/internal/src/generate/clientPreset.ts:35]: It's not related to this PR, it was an extraneous comment that no longer applied and I removed the comment. As can see I removed the `      [`${getPaths().web.base}/types/types.d.ts`]: ` which I thought was needed but realized that preset in fact created the proper files.

This was just preset cleanup -- I guess I could have made a separate PR but as I gather I simply combined it as notice no longer needed.

**Review by Tobbe — APPROVED** [review]
